Close #60 (Switch to released version of serialise)

Mikolaj opened the pull request:
This is WIP. Please don't merge yet. See #60.

adamgundry commented:
It should now be possible to get rid of the Travis CI hacks to use a particular `binary-serialise-cbor` commit (see https://github.com/iconnect/api-tools/blob/master/.travis.yml#L23-L28). Updating the `.travis.yml` more generally would be no bad thing, but is unrelated to this issue.

Mikolaj commented:
Travis is green, but there was much guesswork involved. :)

adamgundry commented on src/Data/API/JSONToCBOR.hs:
```diff
@@ -99,7 +104,18 @@ jsonToCBORBasic bt v = case (bt, v) of
+    (BTutc   , String t) ->
+      TTagged 1000 (TMap [ (TInt 1, TInt secs)
+                         , (TInt (-12), TInt psecs) ])
+        where  -- taken from @Codec.Serialise.Class@:
```
It's a shame we have to duplicate this logic, but I think it's okay for the time being. It might be worth exploring going to `Encoding` rather than `Term` in these functions, which should allow us to reuse `encode` for `UTCTIme`, but that's probably a bit big a change to make right now.

adamgundry commented on src/Data/API/Tools/CBOR.hs:
```diff
@@ -170,15 +170,15 @@ encodeUTCTime :: UTCTime -> Encoding
 encodeUTCTime = encodeUTCTime'
 #else
-encodeUTCTime t = encodeTag 0 <> encode (formatTime defaultTimeLocale fmt t)
+encodeUTCTime t = encode (parseUTC_ $ formatTime defaultTimeLocale fmt t)
```
I don't buy this. It uses `encode` at type `Maybe UTCTime`, which doesn't seem right.

How about we drop support for `aeson < 0.8.1`, get rid of all this complexity and the CPP conditionals?

Mikolaj replied:
Right, forgot to deal with the Maybe and types didn't help this one time. :) Will rip out the old aeson shims instead.

adamgundry commented on src/Data/API/Tools/CBOR.hs:
```diff
@@ -170,15 +170,15 @@ encodeUTCTime :: UTCTime -> Encoding
 -- api-tools, which never include fractions of a second.
 encodeUTCTime' :: UTCTime -> Encoding
-encodeUTCTime' t = encodeTag 0 <> encode (mkUTC' t)
+encodeUTCTime' = encode
```
The comment here is now rather misleading. We're now CBOR-encoding `UTCTime` "normally" and dealing with the correspondence with JSON elsewhere. I'd just inline and get rid of `encodeUTCTime'`.

adamgundry commented on src/Data/API/Value.hs:
```diff
@@ -85,7 +87,7 @@ import           Prelude
 data Value = String  !T.Text
-           | UTCTime !T.Text
+           | UTCTime !UTCTime
              -- ^ A time represented as a string, not decoded for efficiency
```
This comment needs an update. :wink:

adamgundry commented on src/Data/API/Value.hs:
```diff
@@ -310,7 +311,7 @@ matchesNormAPIBasic bt v p = case (bt, v) of
     (BTint, _)           -> Left (JSONError (expectedInt js_v), p)
     (BTutc, UTCTime _)   -> return ()
-    (BTutc, _)           -> Left (JSONError (Expected ExpString "UTCTime" js_v), p)
+    (BTutc, _)           -> Left (JSONError (Expected ExpString "UTCTime" js_v), p)  -- TODO: no @ExpUTCTime@ defined elsewhere, so error not accurate
```
I wouldn't worry too much about precision of these errors, I think this is fine.

adamgundry commented on stack-7.0.yaml:
```diff
@@ -6,10 +6,6 @@ flags:
-      commit: 42af834eb84da3fbb7319a1f6480367ca99dcbfd
 
 extra-deps:
-  - binary-serialise-cbor-0.1.1.0
```
Presumably we ought to specify `serialise`/`cborg` as `extra-deps`.

cdornan commented:
@adamgundry I have no objections at all to dropping support for old packages and simplifying.

Mikolaj commented:
@adamgundry: thank you, I've addressed your comments and, in particular, gutted out the shims for old aeson and related. I agree it would be good to use it with atlas after merge, but before publishing to Hackage.

adamgundry commented:
Thanks, this is coming together nicely. Two final points:
 * I disagree with the characterisation of `Data.API.Tools.Traversal` as unused: it certainly is used downstream!
 * Let's do a major version bump to `0.8.0.0`, as we are making non-backwards-compatible changes.

Mikolaj commented:
@adamgundry: oh, where is `Data.API.Tools.Traversal` used? I searched in both atlas and hermes --- where should I look next time?

Agreed about `0.8.0.0`.

adamgundry commented:
> oh, where is `Data.API.Tools.Traversal` used? I searched in both atlas and hermes --- where should I look next time?

@Mikolaj it's in `Atlas.Types.Traversal` for some cheap generic programming.

Mikolaj commented:
@adamgundry: wow, how did I miss it? Reverting right now. Thank you!
